designs/0004-closures-fun-types.md

```diff
@@ -0,0 +1,608 @@
+the captures at the front of the lambda expression be `[&]`, which
+specifies that all automatic variables used in the body of the lambda
+be captured by reference.  Automatic variables are those without
+explicit capture declarations; all variables in this proposal for Stan
+will behave as automatic variables.
+
+#### R
+
+In R, the expression `function(u) { return(1 / (1 + exp(-u))) }`
+defines a function that denotes the inverse logit function.  It can be
+assigned to a variable, e.g.,
+
+```
+inv_logit <- function(u) { return(1 / (1 + exp(-u))); }
+theta <- inv_logit(0.2)
+```
+
+Unlike C++ or Stan, R requires the outer parentheses for returns.  R
+provides a convenient shortcut that the body of a function returns its
+last evaluated expression if there is no return statement, so the
+inverse logit function can be rewritten as
+
+```
+inv_logit <- function(u) { 1 / (1 + exp(-u)) }
+```
+
+We can go one step further and drop the braces, as they are only there
+to allow a sequence of statements to be grouped,
+
+```
+inv_logit <- function(u) 1 / (1 + exp(-u))
+```
+
+This is the abbreviated syntax we recommend here for Stan lambdas,
+where the equivalent would be
+
+```
+real(real) ilogit = (real u) { return 1 / (1 + exp(-u)); };
+```
+
+or in abbreviated form,
+
+```
+real(real) ilogit = (real u) 1 / (1 + exp(-u));
+```
+
+R uses the unusual mechanism of dynamic lexical scoping, meaning that
+a lambda will capture whichever variable exists in its environment
+when executed.  This can lead to non-deterministic behavior of
+variable scopes at runtime.  This proposal for Stan is to use the more
+traditional approach of static lexical scoping, which is what is used
+in C++.
+
+#### Python
+
+Python allows lambdas such as `lambda u : 1 / (1 + exp(-u))`.
+Multiple argument functions my be `lambda x, y : (x**2 + y**2)**0.5`.
+Python captures variables by reference in lambdas, as the following
+example demonstrates.
+
+```
+>>> x = 3
+>>> f = lambda y : (x**2 + y**2)**0.5
+>>> f(4)
+5.0
+>>> x = 10
+>>> f(4)
+10.770329614269007
+```
+
+This is the same capture-by reference that is used by R and this
+proposal for Stan, as well as the behavior for C++ if the captures
+specification takes all implicit captures by reference (`[&]`).
+
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+1.  Will Stan permit potentially risky capture of local variables and
+    function arguments?  If it is allowed (as it presumably will be,
+    as it's very very limiting not to) is there a way to flag cases
+    that might be potentially dangerous and provide warnings?  
+
+2.  What should the syntax look like?  This proposal just follows the
+	C++ syntax for both types and closures.
+
+3.  Should variables be captured by value instead of reference? That
+    avoids them changing later and avoids dangling references, but it
+    also requires copies and is counterintuitive for those used to
+    R or Python.
```

Comment:
My feeling is that the copies would be so bad that we should not go with by reference. It also feels more consistent with ordinary function argument which are passed by constant reference. Again, I feel like the enclosing variables should be passed by *constant* reference.

Reply:
We can avoid copies in cases where people don't mutate the variables in question in between calls to the lambda. I would say that is 99% of the use cases I've seen.

I don't think people can tell that functions are applied with constant reference vs by value in Stan currently.